hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineManager.java

```diff
@@ -104,12 +104,13 @@ void initializePipelines(List<Pipeline> pipelinesFromScm)
       List<Pipeline> pipelinesInHouse = getPipelines();
       LOG.info("Recon has {} pipelines in house.", pipelinesInHouse.size());
       for (Pipeline pipeline : pipelinesFromScm) {
-        if (!containsPipeline(pipeline.getId())) {
-          // New pipeline got from SCM. Recon does not know anything about it,
-          // so let's add it.
-          LOG.info("Adding new pipeline {} from SCM.", pipeline.getId());
-          addPipeline(pipeline);
+        // New pipeline got from SCM. Recon does not know anything about it,
+        // so let's add it.
+        boolean added = addPipeline(pipeline);
+        if (added) {
+          LOG.info("Added new pipeline {} from SCM.", pipeline.getId());
         } else {
+          LOG.warn("Pipeline {} already exists in Recon pipeline metadata.", pipeline.getId());
           // Recon already has this pipeline. Just update state and creation
           // time.
           getStateManager().updatePipelineState(
```

Comment:
`removeInvalidPipelines(pipelinesFromScm)` this call seems incorrect to be in for loop, though it is old code, because we are removing full list from SCM in every iteration. Also pls check this method is acquiring lock even when caller of this already acquired lock and this method is just called from a single caller.